src/network/networkd-address.c

```diff
@@ -134,8 +137,10 @@ void address_free(Address *address) {
                 set_remove(address->link->dhcp6_addresses_old, address);
                 set_remove(address->link->dhcp6_pd_addresses, address);
                 set_remove(address->link->dhcp6_pd_addresses_old, address);
-                set_remove(address->link->ndisc_addresses, address);
-                set_remove(address->link->ndisc_addresses_old, address);
+                HASHMAP_FOREACH(addresses, address->link->ndisc_addresses_by_router, i)
+                        set_remove(addresses, address);
+                HASHMAP_FOREACH(addresses, address->link->ndisc_addresses_old_by_router, i)
+                        set_remove(addresses, address);
```

Comment:
Hmm... maybe instead of building two parallel hashmap structures, it would be easier to mark the items in those structures with a boolean flag `is_old`? This would require extending the item type, but would otherwise make the code a lot simpler.